ldap/servers/slapd/back-ldbm/db-mdb/mdb_instance.c

```diff
@@ -1581,14 +1590,77 @@ dbmdb_privdb_destroy(mdb_privdb_t **db)
     }
 }
 
+
+/* Compute a smaller key for long rdn */
+int
+dbmdb_privdb_init_small_key(mdb_privdb_t *db, const MDB_val *longkey, int addnewkey, privdb_small_key_t *smallkey)
+{
+    MDB_val key = {0};
+    MDB_val data = {0};
+    unsigned long nextid = 0;
+    char *pt = longkey->mv_data;
+    int rc = 0;
+
+    memset(smallkey, 0, sizeof *smallkey);
+    /* Prepare the samllkey used to store the long key with id=0 */
+    strcpy(smallkey->header, "==>");
+    for (pt += longkey->mv_size - 1; pt >= (char*)longkey->mv_data; pt--) {
+        smallkey->h += ((smallkey->h << 3) | (smallkey->h >> ((8 * sizeof(long)) - 3))) ^ (*pt & 0x1f);
+    }
+    key.mv_data = smallkey;
+    key.mv_size = sizeof (*smallkey);
+    rc = MDB_CURSOR_GET(db->cursor, &key, &data, MDB_SET_RANGE);
+    /* Let walk all the keys that have the right hash */
+    while (rc == 0 && key.mv_size == sizeof(privdb_small_key_t) &&
+           memcmp(key.mv_data, smallkey, offsetof(privdb_small_key_t, id)) == 0) {
+        if (longkey->mv_size == data.mv_size &&
+            memcmp(longkey->mv_data, data.mv_data, data.mv_size) == 0) {
+            /* Key to the wanted data exists, lets return the small key used for getting/storing data ! */
+            memcpy(smallkey, key.mv_data, key.mv_size);
+            smallkey->header[0] = '@';
+            return 0;
+        }
+        /* Wrong key, lets get next one */
+        memcpy(&nextid, &((privdb_small_key_t*)data.mv_data)->id, sizeof nextid);
+        nextid++;
+        rc = MDB_CURSOR_GET(db->cursor, &key, &data, MDB_NEXT);
+    }
+    if (rc ==0 || rc == MDB_NOTFOUND) {
+        /* Key does not exists */
+        if (addnewkey) {
+            /* Lets add it! */
+            smallkey->id = nextid;
+            key.mv_data = smallkey;
+            key.mv_size = sizeof (*smallkey);
+            rc = MDB_CURSOR_PUT(db->cursor, &key, (MDB_val*)longkey, 0);
+            smallkey->header[0] = '@';
+        } else {
+            rc = MDB_NOTFOUND;
+        }
+    }
+    return rc;
+}
+
+
 int
 dbmdb_privdb_get(mdb_privdb_t *db, int dbi_idx, MDB_val *key, MDB_val *data)
 {
     int rc = dbmdb_privdb_handle_cursor(db, 0);
     data->mv_data = NULL;
     data->mv_size = 0;
     if (!rc) {
-        rc = MDB_CURSOR_GET(db->cursor, key, data, MDB_SET_KEY);
+        if (key->mv_size > db->maxkeysize) {
+            privdb_small_key_t small_key = {0};
+            MDB_val key2 = {0};
+            key2.mv_data = &small_key;
+            key2.mv_size = sizeof small_key;
+            rc = dbmdb_privdb_init_small_key(db, key, 0, &small_key);
+            if (!rc) {
+                rc = MDB_CURSOR_GET(db->cursor, &key2, data, MDB_SET_KEY);
```

Comment:
I am surprise it uses the same db->cursor. Depending on the size of the key, is not it using a different database/cursor. Is db->cursor already set to the appropriate DB ?

Reply:
Note: that there are two different mechanism
  [1] For entryrdn (in this case, AFAIK there are two database)
            Entryrdn:   small-key --> long-data  
                is replaced by:
            Entryrdn:   small-key --> redirect-key:
            @longentryrdn: redirect-key: long-data
  [2] For the import temporary private db (which is the case here) 
             privdb: long-key --> data
                  is replaced by: 
             small-key1 --> data
             small-key2 --> long-key     (smallkey1 and smallkey2 are the same except for the first characters (= versus @) 
So yes there is only 1 db (and 1 single cursor) involved in this case